transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/persistence/SQLiteEventStore.java

```diff
@@ -0,0 +1,444 @@
+      whereClause.append(events.get(i).getId());
+      if (i < events.size() - 1) {
+        whereClause.append(',');
+      }
+    }
+    whereClause.append(')');
+
+    try (Cursor cursor =
+        db.query(
+            "event_metadata",
+            new String[] {"event_id", "name", "value"},
+            whereClause.toString(),
+            null,
+            null,
+            null,
+            null)) {
+      while (cursor.moveToNext()) {
+        long eventId = cursor.getLong(0);
+        Set<Metadata> currentSet = metadataIndex.get(eventId);
+        if (currentSet == null) {
+          currentSet = new HashSet<>();
+          metadataIndex.put(eventId, currentSet);
+        }
+        currentSet.add(new Metadata(cursor.getString(1), cursor.getString(2)));
+      }
+    }
+    return metadataIndex;
+  }
+
+  /** Populate metadata into the events. */
+  private List<PersistedEvent> join(
+      List<PersistedEvent> events, Map<Long, Set<Metadata>> metadataIndex) {
+    ListIterator<PersistedEvent> iterator = events.listIterator();
+    while (iterator.hasNext()) {
+      PersistedEvent current = iterator.next();
+      if (!metadataIndex.containsKey(current.getId())) {
+        continue;
+      }
+      EventInternal.Builder newEvent = current.getEvent().toBuilder();
+
+      for (Metadata metadata : metadataIndex.get(current.getId())) {
+        newEvent.addMetadata(metadata.key, metadata.value);
+      }
+      iterator.set(
+          PersistedEvent.create(current.getId(), current.getBackendName(), newEvent.build()));
+    }
+    return events;
+  }
+
+  private <T> T retryIfDbLocked(
+      long lockTimeoutMs, Producer<T> retriable, Function<Throwable, T> failureHandler) {
+    long startTime = monotonicClock.getTime();
+    do {
+      try {
+        return retriable.produce();
+      } catch (SQLiteDatabaseLockedException ex) {
+        if (monotonicClock.getTime() >= startTime + lockTimeoutMs) {
+          return failureHandler.apply(ex);
+        }
+        SystemClock.sleep(LOCK_RETRY_BACK_OFF);
+      }
+    } while (true);
+  }
+
+  interface Producer<T> {
+    T produce();
+  }
+
+  interface Function<T, U> {
+    U apply(T input);
+  }
+
+  /** Tries to start a transaction until it succeeds or times out. */
+  private void ensureBeginTransaction(SQLiteDatabase db, long lockTimeoutMs) {
+    retryIfDbLocked(
+        lockTimeoutMs,
+        () -> {
+          db.beginTransaction();
+          return null;
+        },
+        ex -> {
+          throw new SynchronizationException("Timed out while trying to acquire the lock.", ex);
+        });
+  }
+
+  @Override
+  public <T> T runCriticalSection(long lockTimeoutMs, CriticalSection<T> criticalSection) {
+    SQLiteDatabase db = getDb();
+    ensureBeginTransaction(db, lockTimeoutMs);
+    try {
+      T result = criticalSection.execute();
+      db.setTransactionSuccessful();
+      return result;
+    } finally {
+      db.endTransaction();
```

Comment:
Does this throw any errors that may result in the transaction not ending and causing others to wait?

Reply:
The only documented failure is if someone is holding a shared lock on the db. It is not possible in our case as we use `exclusive` transactions